internal/scheduler/scheduler.go

```diff
@@ -2,344 +2,660 @@ package scheduler
+						Created: s.now(),
+						Event: &armadaevents.EventSequence_Event_JobRunErrors{
+							JobRunErrors: &armadaevents.JobRunErrors{
+								RunId:  armadaevents.ProtoUuidFromUuid(run.RunID),
+								JobId:  jobId,
+								Errors: []*armadaevents.Error{leaseExpiredError},
+							},
+						},
+					},
+					{
+						Created: s.now(),
+						Event: &armadaevents.EventSequence_Event_JobErrors{
+							JobErrors: &armadaevents.JobErrors{
+								JobId:  jobId,
+								Errors: []*armadaevents.Error{leaseExpiredError},
+							},
+						},
+					},
+				},
+			}
+			events = append(events, es)
+			jobsToDelete = append(jobsToDelete, job.JobId)
+		}
+	}
+	err = s.jobDb.BatchDelete(txn, jobsToDelete)
 	if err != nil {
-		return errors.Wrapf(err, "error creating pulsar producer %s", serverPulsarProducerName)
+		return nil, err
 	}
-	defer producer.Close()
+	return events, nil
+}
+
+// now is a convenience function for generating a pointer to  a time.Time (as required by armadaevents).  It exists
+// because Go won't let you do &s.clock.Now()
+func (s *Scheduler) now() *time.Time {
+	now := s.clock.Now()
+	return &now
+}
 
-	scheduler := NewScheduler(producer, db)
-	scheduler.Run(app.CreateContextWithShutdown())
+// initialise builds the initial job db based on the current database state
+// right now this is quite dim and loads the entire database but in the future
+// we should be  able to make it load active jobs/runs only
+func (s *Scheduler) initialise(ctx context.Context) error {
+	for {
+		_, err := s.syncState(ctx)
+		if err != nil {
+			log.WithError(err).Error("Error initialising")
+		} else {
+			break
+		}
+	}
 	return nil
 }
+
+// ensureDbUpToDate  blocks until that the database state contains all Pulsar messages sent *before* this
+// function was called. This is achieved firstly by publishing messages to Pulsar and then polling the
+// database until all messages have been written.
+func (s *Scheduler) ensureDbUpToDate(ctx context.Context, pollInterval time.Duration) error {
+	var groupId uuid.UUID
+	var numSent uint32
+	var err error
+
+	// Send messages to Pulsar
+	messagesSent := false
+	for !messagesSent {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			numSent, err = s.publisher.PublishMarkers(ctx, groupId)
+			if err != nil {
+				log.WithError(err).Error("Error sending marker messages to pulsar")
+				s.clock.Sleep(pollInterval)
+			} else {
+				messagesSent = true
+			}
+		}
+	}
+
+	// Try to read these messages back from the DB
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		default:
+			numReceived, err := s.jobRepository.CountReceivedPartitions(ctx, groupId)
+			if err != nil {
+				log.WithError(err).Error("Error querying the database  or marker messages")
+			}
+			if numSent == numReceived {
+				log.Infof("Successfully ensured that database state is up to date")
+				return nil
+			}
+			log.Infof("Recevied %d partitions, still waiting on  %d", numSent, numSent-numReceived)
```

Comment:
Nit: recommend breaking out formatting arguments into logrus fields.

Reply:
As above- although I'm less convinced this is a useful stat to expose.